service/dap/server.go

```diff
@@ -2915,16 +2913,17 @@ func (s *Server) resumeOnce(command string, allowNextStateChange chan struct{})
 // due to an error, so the server is ready to receive new requests.
 func (s *Server) runUntilStopAndNotify(command string, allowNextStateChange chan struct{}) {
 	// Clear any manual stop requests that came in before we started running.
-	s.checkAndClearHaltRequested()
-	state, manualHalt, err := s.runUntilStop(command, allowNextStateChange)
+	s.setHaltRequested(false)
+	state, err := s.runUntilStop(command, allowNextStateChange)
 
 	if processExited(state, err) {
 		s.send(&dap.TerminatedEvent{Event: *newEvent("terminated")})
 		return
 	}
 
 	stopReason := s.debugger.StopReason()
-	if manualHalt {
+	// Override the stop reason if there was a manual stop request.
+	if s.checkHaltRequested() {
```

Comment:
Or maybe just like you have high level isRunningCmd, you can have high-level StopReason in the server. And then you can set it when you set isRunningCmd(false) and then retrieve here.

Reply:
I disagree that this is a bug, though I agree that there is likely a better way to handle this. I do think a high level StopReason makes more sense because this information may become useful in other requests as well (like threads request, etc).

I don't think it hurts to overwrite the stop reason with stopManual, if there was a halt requested at a real stop on breakpoint, since these could have occurred simultaneously.

Reply:
It's true that the low-level debugger does indeed choose manual reason over breakpoints if they are hit at the same time. This change artificially expands the collision window. Doesn't this hurt us especially in case of panics with have a lot of useful info added? 

If you feel strongly about this, can we at least add a comment/TODO with an explanation and move the code until after the logging statement, so you can log both the reason before and the fact that you overrode it. Then if this ever causes trouble in the wild, we will have more informative logs.

Reply:
I added a TODO for this. It would definitely be helpful to have a better system for both the stop reason and the goroutine id, but I think this should be done in a separate PR, to avoid complicating this one further.

Reply:
It just occurred to me that there is a simple fix for this. You can just move the override statement to after the switch if stopped.Body.Reason != "exception". If you override the reason for any kind of breakpoint, it doesn't really matter. Either way the right line will be highlighted and nobody will care what reason the UI shows. And if you have a panic, you will preserve all the necessary details. And please do add that logging statement we talked about, so you log both the internal reason and the fact that you changed it.